builtin/logical/pki/backend_test.go

```diff
@@ -2178,9 +2178,10 @@ func runTestSignVerbatim(t *testing.T, keyType string) {
 
 	// create a role entry; we use this to check that sign-verbatim when used with a role is still honoring TTLs
 	roleData := map[string]interface{}{
-		"ttl":      "4h",
-		"max_ttl":  "8h",
-		"key_type": keyType,
+		"ttl":                 "4h",
+		"max_ttl":             "8h",
+		"key_type":            keyType,
+		"not_before_duration": "2h",
```

Comment:
One small test improvement :-) If you look below around line 2250, we test that the requested 12h TTL wasn't honored (and is presumably 8h then).

If you don't mind, it'd be great to add a test to make sure the 2h not before duration was honored though :-) 

I ended up doing something like:

```diff
diff --git a/builtin/logical/pki/backend_test.go b/builtin/logical/pki/backend_test.go
index b1402ac82b..6ddd8b4a70 100644
--- a/builtin/logical/pki/backend_test.go
+++ b/builtin/logical/pki/backend_test.go
@@ -2248,7 +2248,10 @@ func runTestSignVerbatim(t *testing.T, keyType string) {
        }
        cert := certs[0]
        if math.Abs(float64(time.Now().Add(12*time.Hour).Unix()-cert.NotAfter.Unix())) < 10 {
-               t.Fatalf("sign-verbatim did not properly cap validity period on signed CSR")
+               t.Fatalf("sign-verbatim did not properly cap validity period (notAfter) on signed CSR: was %v vs requested %v but should've been %v", cert.NotAfter, time.Now().Add(12*time.Hour), time.Now().Add(8*time.Hour))
+       }
+       if math.Abs(float64(time.Now().Add(-2*time.Hour).Unix()-cert.NotBefore.Unix())) > 10 {
+               t.Fatalf("sign-verbatim did not properly cap validity period (notBefore) on signed CSR: was %v vs expected %v", cert.NotBefore, time.Now().Add(-2*time.Hour))
        }
 
        // Now check signing a certificate using the not_after input using the Y10K value
```

Otherwise, this looks good :-)

Reply:
Well, the PR  was "ready" to be reviewed but still, thanks for feedback.
I think I am a bit confused with these `ttl` and `max_ttl` values. In the first `if` clause we are validating if the `ttl` defined in CSR (12 hours) is not honored, by checking if the difference between `notAfter` value and `now + 12 hours` is more than 10 seconds, right?

My biggest confusion is with the second `if` statement, we want to validate if the 2h `not_before` is honored. Isn't checking that the difference between `notBefore` and `now - 2 hours` is less than 10 seconds what we are looking for?

I have printed the `notAfter` and `notBefore` values of a certificate created now (2022/05/16 17:56:33 in my timezone) and `notAfter` has the value of `now` + 7 hours and `notBefore` now - 2 hours. Without considering the the `notBefore` that's 8 around which is the role `max_ttl`. Does that mean that `notBefore` is not included in the `ttl`? And `notBefore` has always 1 hour by default? These values are based on the test I am reviewing.

I should have known this already but a clarification would be appreciated. Thanks

Reply:
Correct, `notBefore` is not included in the TTL. Max TTL is the maximum allowable TTL the certificate can be valid for (from server time now, to the end of the cert.NotAfter period). TTL is the _default_ TTL the certificate is valid for (again, from server now). 

So, this role says, by default, issue for 4 hours, but let users request up to 8 hour validity (from now).

By allowing a 2 hour backdate, if the servers were somehow that far behind, they could theoretically request up to a 10 hour valid certificate (two in the past for clock sync issues, and 8 more in the future). 

So either way, the requested 12h validity is too much, and is truncated to 8 hours.

---

So, back to your question, we want the success condition to be `(our not_before_duration was respected)`. This expands to something like `(delta (two hours before now) to (cert.NotBefore)) == 0`. However, some of these checks take some amount of time, in, e.g., slow CI systems, so we check `(delta ...) <= 10 seconds` to give some time. Also, we throw in an absolute value for some reason. Maybe clock skew or something, I dunno.

Negating that, gives us `abs(delta) > 10 seconds` as the conditional for which we should abort the test (`t.Fatalf(...)`) on. So I think it is correct?

Reply:
Yes, makes total sense. I was confusing everything in the beginning.

Reply:
If you want, it'd be totally appropriate to add a note to the [API docs](https://github.com/hashicorp/vault/blob/main/website/content/api-docs/secret/pki.mdx#createupdate-role) to notate that `not_before_duration` doesn't impact the `ttl` and that the `ttl` is strictly used for future validity (now to NotAfter).

Reply:
Good idea, I think that would avoid possible confusions from others.
Just pushed a proposal with the additions.